audit/audit.go

```diff
@@ -0,0 +1,164 @@
+package audit
+
+import (
+	"context"
+	"github.com/ONSdigital/go-ns/identity"
+	"github.com/ONSdigital/go-ns/log"
+	"github.com/pkg/errors"
+	"net/http"
+	"time"
+)
+
+//go:generate moq -out generated_mocks.go -pkg audit . AuditorService OutboundProducer
+
+type contextKey string
+
+const (
+	eventContextKey = contextKey("audit")
+	fromContextKey  = contextKey("from")
+	requestIDHeader = "X-Request-Id"
+	fromHeader      = "from"
+)
+
+//Event holds data about the action being attempted
+type Event struct {
+	Created         string         `avro:"created" json:"created,omitempty"`
+	Service         string         `avro:"service" json:"service,omitempty"`
+	Namespace       string         `avro:"namespace" json:"namespace,omitempty"`
+	RequestID       string         `avro:"request_id" json:"request_id,omitempty"`
+	User            string         `avro:"user" json:"user,omitempty"`
+	AttemptedAction string         `avro:"attempted_action" json:"attempted_action,omitempty"`
+	Result          string         `avro:"result" json:"result,omitempty"`
+	Params          []keyValuePair `avro:"params" json:"params,omitempty"`
+}
+
+type avroMarshaller func(s interface{}) ([]byte, error)
+
+//Params key value pair for additional audit data not included in the event model.
+type Params map[string]string
+
+type keyValuePair struct {
+	Key   string `avro:"key" json:"key,omitempty"`
+	Value string `avro:"value" json:"value,omitempty"`
+}
+
+// OutboundProducer defines a producer for sending outbound audit events to a kafka topic
+type OutboundProducer interface {
+	Output() chan []byte
+}
+
+//AuditorService defines the behaviour of an audior
+type AuditorService interface {
+	GetEvent(input context.Context) (*Event, error)
+	Record(ctx context.Context, action string, result string, params Params) error
+}
+
+//Auditor provides functions for interception HTTP requests and populating the context with a base audit event and
+// recording audit events
+type Auditor struct {
+	namespace     string
+	tokenName     string
```

Comment:
This looks like its the header name for service tokens - which is always `Authorization`. Consider removing tokenName from the struct and using the const defined here: https://github.com/ONSdigital/go-ns/blob/4001bc2311e9cd2de98a365b6ff5ca6b1065e576/common/identity.go#L14